File: RANDEVU/backend/llm_service.py
#!/usr/bin/env python3
"""
LLM Servisi
Ollama ve diğer LLM sağlayıcıları ile entegrasyon
"""

import os
import json
import requests
import logging
from typing import Dict, Any, Optional
from datetime import datetime

class LLMService:
    """LLM servisi - Ollama ve diğer sağlayıcılar"""
    
    def __init__(self):
        self.ollama_url = os.getenv("OLLAMA_HOST", "http://localhost:11434")
        self.model_name = os.getenv("LITELLM_MODEL", "llama3.2:3b")
        self.timeout = int(os.getenv("LLM_TIMEOUT", "30"))
        
        # Model sağlayıcısını belirle
        if self.model_name.startswith("ollama/"):
            self.provider = "ollama"
            self.model_name = self.model_name.replace("ollama/", "")
        elif self.model_name.startswith("gpt-"):
            self.provider = "openai"
        elif self.model_name.startswith("gemini/"):
            self.provider = "google"
        else:
            self.provider = "ollama"  # Varsayılan
        
        logging.info(f"LLM Service initialized: {self.provider}/{self.model_name}")
    
    def get_chat_response(self, message: str, context: str = "medical_assistant") -> str:
        """Chat yanıtı al"""
        try:
            if self.provider == "ollama":
                return self._call_ollama(message, context)
            elif self.provider == "openai":
                return self._call_openai(message, context)
            elif self.provider == "google":
                return self._call_google(message, context)
            else:
                return self._get_fallback_response(message)
        except Exception as e:
            logging.error(f"LLM call error: {e}")
            return self._get_fallback_response(message)
    
    def _call_ollama(self, message: str, context: str) -> str:
        """Ollama API'sini çağır"""
        try:
            # Sistem prompt'u oluştur
            system_prompt = self._get_system_prompt(context)
            
            # Prompt'u hazırla
            full_prompt = f"{system_prompt}\n\nKullanıcı: {message}\nAsistan:"
            
            payload = {
                "model": self.model_name,
                "prompt": full_prompt,
                "stream": False,
                "options": {
                    "temperature": 0.1,
                    "max_tokens": 512,
                    "top_p": 0.9,
                    "repeat_penalty": 1.1
                }
            }
            
            response = requests.post(
                f"{self.ollama_url}/api/generate",
                json=payload,
                timeout=self.timeout
            )
            
            if response.status_code == 200:
                result = response.json()
                return result.get('response', '').strip()
            else:
                logging.error(f"Ollama API error: {response.status_code}")
                return self._get_fallback_response(message)
                
        except requests.exceptions.Timeout:
            logging.error("Ollama timeout")
            return self._get_fallback_response(message)
        except Exception as e:
            logging.error(f"Ollama call error: {e}")
            return self._get_fallback_response(message)
    
    def _call_openai(self, message: str, context: str) -> str:
        """OpenAI API'sini çağır"""
        try:
            import openai
            
            client = openai.OpenAI(
                api_key=os.getenv("OPENAI_API_KEY")
            )
            
            system_prompt = self._get_system_prompt(context)
            
            response = client.chat.completions.create(
                model=self.model_name,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": message}
                ],
                temperature=0.1,
                max_tokens=512
            )
            
            return response.choices[0].message.content.strip()
            
        except Exception as e:
            logging.error(f"OpenAI call error: {e}")
            return self._get_fallback_response(message)
    
    def _call_google(self, message: str, context: str) -> str:
        """Google Gemini API'sini çağır"""
        try:
            import google.generativeai as genai
            
            genai.configure(api_key=os.getenv("GOOGLE_API_KEY"))
            
            model = genai.GenerativeModel(self.model_name.replace("gemini/", ""))
            
            system_prompt = self._get_system_prompt(context)
            full_prompt = f"{system_prompt}\n\nKullanıcı: {message}"
            
            response = model.generate_content(
                full_prompt,
                generation_config=genai.types.GenerationConfig(
                    temperature=0.1,
                    max_output_tokens=512
                )
            )
            
            return response.text.strip()
            
        except Exception as e:
            logging.error(f"Google Gemini call error: {e}")
            return self._get_fallback_response(message)
    
    def _get_system_prompt(self, context: str) -> str:
        """Sistem prompt'unu al"""
        if context == "medical_assistant":
            return """Sen TanıAI Asistanısın, Türkiye'de sağlık hizmetleri konusunda uzman bir yapay zeka asistanısın. 

Görevlerin:
1. Kullanıcıların sağlık sorularını yanıtlamak
2. Randevu alma konusunda rehberlik etmek
3. Semptom analizi yapmak (sadece genel bilgi amaçlı)
4. Uygun klinik önerileri sunmak
5. Eczane ve ilaç bilgileri vermek

Önemli kurallar:
- Tıbbi teşhis koyma, sadece genel bilgi ver
- Acil durumlarda 112'yi aramalarını söyle
- Doktor muayenesi öner
- Türkçe yanıt ver
- Kısa ve net yanıtlar ver
- Empatik ve anlayışlı ol"""
        
        return "Sen yardımcı bir AI asistanısın. Kullanıcılara yardımcı olmaya çalış."
    
    def _get_fallback_response(self, message: str) -> str:
        """Fallback yanıtlar"""
        message_lower = message.lower()
        
        if any(word in message_lower for word in ['randevu', 'appointment', 'doktor']):
            return "Randevu almak için size yardımcı olabilirim! Hangi bölüm için randevu almak istiyorsunuz?"
        
        elif any(word in message_lower for word in ['semptom', 'ağrı', 'hasta']):
            return "Semptomlarınız hakkında konuşabiliriz, ancak kesin teşhis için mutlaka bir doktora başvurmanızı öneririm."
        
        elif any(word in message_lower for word in ['eczane', 'ilaç']):
            return "Eczane ve ilaç konularında size yardımcı olabilirim. Yakınızdaki eczaneleri bulmak için eczane bulma özelliğini kullanabilirsiniz."
        
        elif any(word in message_lower for word in ['merhaba', 'selam', 'hello']):
            return "Merhaba! Ben TanıAI Asistanıyım. Size nasıl yardımcı olabilirim?"
        
        else:
            return "Anlıyorum. Size daha iyi yardımcı olabilmem için daha spesifik bir soru sorabilir misiniz?"

File: RANDEVU/mobile_flutter/lib/services/llm_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class LLMService {
  static const String _baseUrl = 'http://localhost:8001'; // FastAPI backend
  static const Duration _timeout = Duration(seconds: 30);

  /// LLM ile chat yanıtı al
  static Future<String> getChatResponse(String userMessage) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/chat'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'message': userMessage,
          'context': 'medical_assistant',
        }),
      ).timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['response'] ?? 'Üzgünüm, bir yanıt oluşturamadım.';
      } else {
        return 'Üzgünüm, şu anda bir sorun yaşıyorum. Lütfen daha sonra tekrar deneyin.';
      }
    } catch (e) {
      print('LLM API Error: $e');
      return _getFallbackResponse(userMessage);
    }
  }

  /// Klinik triyaj için LLM yanıtı al
  static Future<Map<String, dynamic>> getTriageResponse(String symptoms) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/triage'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'text': symptoms,
        }),
      ).timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return {
          'success': true,
          'data': data,
        };
      } else {
        return {
          'success': false,
          'error': 'Triyaj servisi şu anda kullanılamıyor.',
        };
      }
    } catch (e) {
      print('Triage API Error: $e');
      return {
        'success': false,
        'error': 'Triyaj servisi şu anda kullanılamıyor.',
      };
    }
  }

  /// Klinik önerisi al
  static Future<Map<String, dynamic>> getClinicRecommendation(String symptoms) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/recommend-clinic'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'symptoms': symptoms,
        }),
      ).timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return {
          'success': true,
          'data': data,
        };
      } else {
        return {
          'success': false,
          'error': 'Klinik önerisi servisi şu anda kullanılamıyor.',
        };
      }
    } catch (e) {
      print('Clinic Recommendation API Error: $e');
      return {
        'success': false,
        'error': 'Klinik önerisi servisi şu anda kullanılamıyor.',
      };
    }
  }

  /// Fallback yanıtlar (LLM servisi çalışmıyorsa)
  static String _getFallbackResponse(String userMessage) {
    final message = userMessage.toLowerCase();
    
    // Baş ağrısı - spesifik yanıt
    if (message.contains('baş') && message.contains('ağrı')) {
      return 'Baş ağrınız için size yardımcı olabilirim. Baş ağrısı birçok nedenden kaynaklanabilir:\n\n'
          '🔍 **Olası Nedenler:**\n'
          '• Stres ve yorgunluk\n'
          '• Migren\n'
          '• Sinüzit\n'
          '• Tansiyon\n'
          '• Dehidrasyon\n\n'
          '💡 **Öneriler:**\n'
          '• Bol su için\n'
          '• Karanlık ve sessiz bir yerde dinlenin\n'
          '• Hafif masaj yapın\n'
          '• Ağrı kesici alabilirsiniz\n\n'
          '⚠️ **Dikkat:** Şiddetli, ani başlayan veya sürekli baş ağrıları için mutlaka doktora başvurun.';
    }
    
    // Karın ağrısı - spesifik yanıt
    if (message.contains('karın') && message.contains('ağrı')) {
      return 'Karın ağrınız için size yardımcı olabilirim. Karın ağrısı farklı nedenlerden kaynaklanabilir:\n\n'
          '🔍 **Olası Nedenler:**\n'
          '• Hazımsızlık\n'
          '• Gaz\n'
          '• Stres\n'
          '• Gıda intoleransı\n'
          '• Mide ülseri\n\n'
          '💡 **Öneriler:**\n'
          '• Hafif yiyecekler tüketin\n'
          '• Bol su için\n'
          '• Sıcak su torbası kullanın\n'
          '• Dinlenin\n\n'
          '⚠️ **Dikkat:** Şiddetli, ani başlayan veya sürekli karın ağrıları için mutlaka doktora başvurun.';
    }
    
    // Göğüs ağrısı - acil durum
    if (message.contains('göğüs') && message.contains('ağrı')) {
      return '🚨 **ACİL DURUM!**\n\n'
          'Göğüs ağrısı ciddi bir durum olabilir. Hemen aşağıdaki adımları takip edin:\n\n'
          '📞 **Hemen Yapılacaklar:**\n'
          '• 112\'yi arayın\n'
          '• Acil servise gidin\n'
          '• Yanınızda birisi olsun\n'
          '• Dinlenin, hareket etmeyin\n\n'
          '⚠️ **Göğüs ağrısı kalp krizi belirtisi olabilir. Zaman kaybetmeyin!**';
    }
    
    // Randevu ile ilgili
    if (message.contains('randevu') || message.contains('appointment') || message.contains('doktor')) {
      return 'Randevu almak için size yardımcı olabilirim! 🏥\n\n'
          '**Hangi bölüm için randevu almak istiyorsunuz?**\n\n'
          '📋 **Mevcut Bölümler:**\n'
          '• Genel Pratisyen\n'
          '• Kardiyoloji\n'
          '• Nöroloji\n'
          '• Ortopedi\n'
          '• Dermatoloji\n'
          '• İç Hastalıkları\n\n'
          '💡 **Randevu Seçenekleri:**\n'
          '• 📞 Sesli randevu (telefon)\n'
          '• 🏥 Yakın hastaneler\n'
          '• 📅 Online randevu sistemi';
    }
    
    // Semptom ile ilgili - genel
    if (message.contains('semptom') || message.contains('hasta') || message.contains('ağrı') || message.contains('rahatsızlık')) {
      return 'Semptomlarınız hakkında konuşabiliriz. Size daha iyi yardımcı olabilmem için:\n\n'
          '🔍 **Lütfen belirtin:**\n'
          '• Hangi semptomları yaşıyorsunuz?\n'
          '• Ne kadar süredir devam ediyor?\n'
          '• Şiddeti nasıl? (hafif/orta/şiddetli)\n'
          '• Başka belirtiler var mı?\n\n'
          '💡 **Örnek semptomlar:**\n'
          '• Baş ağrısı\n'
          '• Karın ağrısı\n'
          '• Ateş\n'
          '• Bulantı\n'
          '• Yorgunluk\n\n'
          '⚠️ **Önemli:** Bu analiz sadece genel bilgi amaçlıdır. Kesin teşhis için mutlaka bir doktora başvurun.';
    }
    
    // Eczane ile ilgili
    if (message.contains('eczane') || message.contains('ilaç') || message.contains('pharmacy')) {
      return 'Eczane ve ilaç konularında size yardımcı olabilirim! 💊\n\n'
          '🏥 **Eczane Hizmetleri:**\n'
          '• Yakınızdaki eczaneleri bulun\n'
          '• Nöbetçi eczaneleri görün\n'
          '• İlaç bilgileri alın\n'
          '• Reçete sorgulama\n\n'
          '📱 **Nasıl Kullanabilirsiniz:**\n'
          '• "Eczane bul" butonuna tıklayın\n'
          '• Konumunuzu paylaşın\n'
          '• En yakın eczaneleri görün\n\n'
          '💡 **İpucu:** Nöbetçi eczane için 114\'ü arayabilirsiniz.';
    }
    
    // Selamlama
    if (message.contains('merhaba') || message.contains('selam') || message.contains('hello') || message.contains('hi')) {
      return 'Merhaba! 👋 Ben TanıAI Asistanıyım.\n\n'
          '🏥 **Size nasıl yardımcı olabilirim?**\n\n'
          '📋 **Hizmetlerim:**\n'
          '• 🩺 Semptom analizi\n'
          '• 📅 Randevu alma\n'
          '• 💊 Eczane bulma\n'
          '• 🏥 Klinik önerisi\n'
          '• 📊 Sağlık bilgileri\n\n'
          '💬 **Sadece konuşarak başlayın:**\n'
          '"Başım ağrıyor" veya "Randevu almak istiyorum" gibi.';
    }
    
    // Teşekkür
    if (message.contains('teşekkür') || message.contains('sağol') || message.contains('thanks')) {
      return 'Rica ederim! 😊\n\n'
          'Size yardımcı olabildiğim için mutluyum. Sağlığınız için her zaman buradayım.\n\n'
          '💡 **Başka bir konuda yardıma ihtiyacınız olursa çekinmeden sorun!**';
    }
    
    // Veda
    if (message.contains('görüşürüz') || message.contains('bye') || message.contains('hoşça kal')) {
      return 'Hoşça kalın! 👋\n\n'
          'Sağlığınız için her zaman buradayım. İyi günler dilerim! 🌟\n\n'
          '💡 **Unutmayın:** Sağlık sorunlarınız için mutlaka doktora başvurun.';
    }
    
    // Genel yanıt - daha yardımcı
    return 'Anlıyorum. Size daha iyi yardımcı olabilmem için daha spesifik bilgi verebilir misiniz? 🤔\n\n'
        '💡 **Örnek sorular:**\n'
        '• "Başım ağrıyor"\n'
        '• "Randevu almak istiyorum"\n'
        '• "Yakınımdaki eczaneleri bul"\n'
        '• "Karın ağrım var"\n\n'
        '🏥 **Hizmetlerim:**\n'
        '• Semptom analizi\n'
        '• Randevu yönlendirmesi\n'
        '• Eczane bulma\n'
        '• Klinik önerisi';
  }
}
